src/app/core/store/documents/documents.effects.ts

```diff
@@ -292,53 +278,50 @@ export class DocumentsEffects {
 }
 
 
-function saveNewAttributes(collection: CollectionModel,
-                           documentDto: Document,
+function saveNewAttributes(collectionId: string,
+                           document: DocumentModel,
                            collectionService: CollectionService): Observable<{ documentDto: Document, attributes: AttributeModel[] }> {
-
-  const newAttributes = findNewAttributes(collection, documentDto)
-    .map(attr => CollectionConverter.toAttributeDto(attr));
+  const newAttributes = document.data.filter(d => isNullOrUndefined(d.attributeId))
+    .map(d => ({name: d.name, constraints: [], usageCount: 0}));
   if (newAttributes.length === 0) {
+
+    const documentDto = DocumentConverter.toDto(document);
     return Observable.of({documentDto, attributes: newAttributes});
   } else {
-    return collectionService.createAttributes(documentDto.collectionId, newAttributes).pipe(
-      map(attributes => ({documentDto, attributes}))
-    )
+    return collectionService.createAttributes(collectionId, newAttributes).pipe(
+      map(attributes => {
+        const data = document.data.map(d => ({...d, attributeId: d.attributeId || attributes.find(attr => attr.name === d.name).id}));
+        const attributesModels = attributes.map(attr => CollectionConverter.fromAttributeDto(attr));
+        return {documentDto: DocumentConverter.toDto({...document, data}), attributes: attributesModels};
+      })
+    );
   }
 }
 
-function findNewAttributes(collection: CollectionModel, newDocument: Document): AttributeModel[] {
-  const newAttributeNames: string[] = newDocument && newDocument.data ? Object.keys(newDocument.data) : [];
-
-  const attributeNames = collection.attributes.map(attribute => attribute.name);
-  return newAttributeNames.filter(name => !attributeNames.includes(name))
-    .map(name => ({name: name, constraints: []}));
-}
-
 function createSyncCollectionAction(collection: CollectionModel,
                                     newDocument: DocumentModel,
                                     oldDocument: DocumentModel): CollectionsAction.UpdateSuccess {
-  const newAttributeNames: string[] = newDocument && newDocument.data ? Object.keys(newDocument.data) : [];
-  const oldAttributeNames: string[] = oldDocument && oldDocument.data ? Object.keys(oldDocument.data) : [];
+  const newAttributeIds: string[] = newDocument && newDocument.data ? newDocument.data.map(d => d.attributeId) : [];
+  const oldAttributeIds: string[] = oldDocument && oldDocument.data ? oldDocument.data.map(d => d.attributeId) : [];
```

Comment:
Do we still need this if all attribute definitions must exist before storing their values?

Reply:
This is just for increasing/decreasing attributes count in collection